File: web/gpv2.js
// GPv2 settlement decoding + surplus primitives, in the browser.
// Hand-rolled minimal ABI decoder for the fixed settle() signature — no
// dependencies, so the page stays tiny and auditable. Validated to decode
// byte-identically to the Python cow_certify.gpv2 across the full corpus.
//
//   settle(address[] tokens, uint256[] clearingPrices,
//          Trade[] trades, Interaction[][3] interactions)
//
// The reader is STRICT: every word is bounds-checked (a short read throws
// instead of decoding as zero — a zero-filled read once fabricated a signed
// limit of 0 and certified a limit-breaking trade as PASS), and padding bytes
// are validated the way eth_abi's strict decoder validates them. Whether the
// blob is laid out canonically is tracked while decoding, so the appended
// auction id is read only from a canonical blob with EXACTLY an 8-byte tail —
// mirroring the Python decoder's re-encode-and-compare rule.

export const SETTLEMENT = "0x9008d19f58aabd9ed0d60971565aa8510560ab41";
export const SETTLE_SELECTOR = "0x13d79a0b";
export const TRADE_TOPIC =
  "0xa07a543ab8a018198e99ca0184c93fe9050a79400a0a723441f84de1d972cc17";
// Settlement(address indexed solver) — emitted once per settle() even with no
// user Trade events; proof the tx is a real CoW settlement. keccak Settlement(address).
export const SETTLEMENT_EVENT_TOPIC =
  "0x40338ce1a7c49204f0099533b1e9a7ee0a3d261f84974ab7af36105b8c4e9db4";
export const ERC20_TRANSFER =
  "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";
export const AUCTION_ID_SUFFIX_BYTES = 8;
export const FLAG_KIND_BUY = 1n;
export const FLAG_PARTIALLY_FILLABLE = 2n;

export function ceilDiv(a, b) {
  return (a + b - 1n) / b;
}

const pad32 = (n) => Math.ceil(n / 32) * 32;
const MAX_SAFE = BigInt(Number.MAX_SAFE_INTEGER);

// --- minimal, strict ABI reader over the calldata after the 4-byte selector ---
function reader(calldataHex) {
  const hex = calldataHex.startsWith("0x") ? calldataHex.slice(2) : calldataHex;
  const data = hex.slice(8); // drop selector
  const word = (byteOff) => {
    const a = byteOff * 2;
    if (byteOff < 0 || a + 64 > data.length)
      throw new Error(`calldata truncated: word at byte ${byteOff} lies beyond the ${data.length / 2} bytes present`);
    return data.slice(a, a + 64);
  };
  const int = (byteOff) => {
    const v = BigInt("0x" + word(byteOff));
    if (v > MAX_SAFE) throw new Error("offset or length exceeds calldata");
    return Number(v);
  };
  return {
    hex,
    dataBytes: data.length / 2,
    uint: (byteOff) => BigInt("0x" + word(byteOff)),
    int,
    u32: (byteOff) => {
      const w = word(byteOff);
      if (!/^0{56}/.test(w)) throw new Error("malformed uint32 word (non-zero padding)");
      return Number(BigInt("0x" + w));
    },
    addr: (byteOff) => {
      const w = word(byteOff);
      if (!/^0{24}/.test(w)) throw new Error("malformed address word (non-zero padding)");
      return ("0x" + w.slice(24)).toLowerCase();
    },
    word32: (byteOff) => "0x" + word(byteOff),
    // dynamic `bytes` at byteOff: [length][data padded to 32]; returns the hex
    // of the data and validates that the padding is present and zero.
    bytesAt: (byteOff) => {
      const len = int(byteOff);
      const a = (byteOff + 32) * 2;
      const paddedLen = pad32(len) * 2;
      if (a + paddedLen > data.length) throw new Error("calldata truncated: bytes lie beyond the data present");
      const padding = data.slice(a + len * 2, a + paddedLen);
      if (!/^0*$/.test(padding)) throw new Error("malformed bytes (non-zero padding)");
      return data.slice(a, a + len * 2);
    },
  };
}

export function decodeSettlement(calldataHex) {
  if (!/^0x/i.test(calldataHex)) calldataHex = "0x" + calldataHex;
  if (calldataHex.slice(0, 10).toLowerCase() !== SETTLE_SELECTOR) {
    throw new Error("not a GPv2 settle() call");
  }
  const r = reader(calldataHex);
  // `canonical` records whether every dynamic offset equals the one a canonical
  // encoder would have written. A non-canonical blob still decodes (the EVM
  // reads it the same way) but carries no readable autopilot suffix.
  let canonical = true;
  const expect = (actual, want) => { if (actual !== want) canonical = false; };

  const offTokens = r.int(0), offPrices = r.int(32),
        offTrades = r.int(64), offInter = r.int(96);

  // A malicious/garbage RPC response can declare an array length of 2^256-1;
  // looping to it would freeze the tab (no worker to kill). Bound every length
  // by the words the calldata actually contains. (Python's eth_abi raises on
  // the same input in ~1ms.)
  const maxWords = r.dataBytes / 32;
  const len = (off) => {
    const n = r.int(off);
    if (n < 0 || n > maxWords) throw new Error("array length exceeds calldata");
    return n;
  };

  let cursor = 128; // canonical position right after the 4 head words
  expect(offTokens, cursor);
  const nTokens = len(offTokens);
  const tokens = [];
  for (let i = 0; i < nTokens; i++)
    tokens.push(r.addr(offTokens + 32 + i * 32));
  cursor += 32 + 32 * nTokens;

  expect(offPrices, cursor);
  const nPrices = len(offPrices);
  const clearingPrices = [];
  for (let i = 0; i < nPrices; i++)
    clearingPrices.push(r.uint(offPrices + 32 + i * 32));
  cursor += 32 + 32 * nPrices;

  // Trade[]: dynamic array of dynamic tuples (the 11th field, `bytes
  // signature`, makes the tuple dynamic). Element offsets are relative to the
  // position after the length word.
  expect(offTrades, cursor);
  const trades = [];
  const tbase = offTrades + 32;
  const nTrades = len(offTrades);
  let rel = 32 * nTrades; // canonical relative offset of the first tuple
  for (let i = 0; i < nTrades; i++) {
    const relOff = r.int(tbase + i * 32);
    expect(relOff, rel);
    const s = tbase + relOff;
    trades.push({
      sellTokenIndex: r.int(s),
      buyTokenIndex: r.int(s + 32),
      receiver: r.addr(s + 64),
      sellAmount: r.uint(s + 96),
      buyAmount: r.uint(s + 128),
      validTo: r.u32(s + 160),
      appData: r.word32(s + 192),
      feeAmount: r.uint(s + 224),
      flags: r.uint(s + 256),
      executedAmount: r.uint(s + 288),
    });
    const sigOff = r.int(s + 320); // offset of `signature` relative to the tuple
    expect(sigOff, 352);
    const sig = r.bytesAt(s + sigOff);
    rel += 352 + 32 + pad32(sig.length / 2);
  }
  cursor = tbase + rel;

  // Interaction[][3]: fixed-3 array of dynamic arrays of dynamic tuples.
  expect(offInter, cursor);
  const interactions = [];
  let irel = 96; // three stage offsets first
  for (let stage = 0; stage < 3; stage++) {
    const so = r.int(offInter + stage * 32);
    expect(so, irel);
    const sp = offInter + so;
    const nInter = len(sp);
    const sbase = sp + 32;
    let jrel = 32 * nInter;
    const list = [];
    for (let j = 0; j < nInter; j++) {
      const jo = r.int(sbase + j * 32);
      expect(jo, jrel);
      const it = sbase + jo;
      const cdOff = r.int(it + 64);
      expect(cdOff, 96);
      const cd = r.bytesAt(it + cdOff);
      list.push({
        target: r.addr(it),
        value: r.uint(it + 32),
        selector: cd.length >= 8 ? "0x" + cd.slice(0, 8) : "0x",
      });
      jrel += 96 + 32 + pad32(cd.length / 2);
    }
    interactions.push(list);
    irel += 32 + jrel;
  }
  cursor = offInter + irel;

  // Auction id: the autopilot appends EXACTLY 8 bytes to a canonical blob. Any
  // other tail — none, or a non-standard length a custom driver could have
  // appended — is not an autopilot suffix and yields null, never a guessed id.
  let suffixLen = r.dataBytes - cursor;
  if (suffixLen < 0) { canonical = false; suffixLen = 0; }
  const auctionId = (canonical && suffixLen === AUCTION_ID_SUFFIX_BYTES)
    ? BigInt("0x" + r.hex.slice(-2 * AUCTION_ID_SUFFIX_BYTES)) : null;

  return { tokens, clearingPrices, trades, interactions, auctionId, suffixLen, canonical };
}

export function tradeEvents(logs) {
  const out = [];
  for (const lg of logs || []) {
    const topics = lg.topics || [];
    if (!topics.length || topics[0].toLowerCase() !== TRADE_TOPIC) continue;
    if ((lg.address || "").toLowerCase() !== SETTLEMENT) continue;
    const d = (lg.data || "").slice(2);
    if (d.length < 560) continue; // 5 words + offset + length + 56-byte uid
    out.push({
      owner: ("0x" + topics[1].slice(-40)).toLowerCase(),
      sell_token: ("0x" + d.slice(24, 64)).toLowerCase(),
      buy_token: ("0x" + d.slice(64 + 24, 128)).toLowerCase(),
      sell_amount: BigInt("0x" + d.slice(128, 192)),
      buy_amount: BigInt("0x" + d.slice(192, 256)),
      fee_amount: BigInt("0x" + d.slice(256, 320)),
      uid: ("0x" + d.slice(448, 560)).toLowerCase(),
    });
  }
  return out;
}

// Before-fee surplus of one trade, in SURPLUS-TOKEN atoms, from the EXECUTED
// Trade-event amounts against the signed limit — no clearing price (the
// un-forgeable form; mirrors Python gpv2.surplus_from_execution). execSellNet
// is the executed sell amount net of fee.
export function surplusFromExecution(kind, execSellNet, execBuy, limitSell, limitBuy) {
  if (!(limitSell && limitBuy && execSellNet && execBuy)) return 0n;
  if (kind === "sell") {
    return execBuy - ceilDiv(execSellNet * limitBuy, limitSell);
  }
  return (execBuy * limitSell) / limitBuy - execSellNet;
}

File: cow_certify/gpv2.py
"""GPv2 settlement decoding + surplus primitives.

Vendored from cow-backtester (github.com/KaiserSolver/cow-backtester, MIT, same
author) so cow-certify is self-contained and installable anywhere — no external
checkout on the path. The only third-party dependency is `eth_abi` for ABI
decoding of the settle() calldata.

Kept identical to the source so certificates reconstruct settlements the same
way the backtester does; `selftest()` pins the surplus math.
"""
from eth_abi import decode, encode

# GPv2Settlement — same address on every CoW chain.
SETTLEMENT = "0x9008d19f58aabd9ed0d60971565aa8510560ab41"
SETTLE_SELECTOR = "0x13d79a0b"
# Trade(address indexed owner, address sellToken, address buyToken,
#       uint256 sellAmount, uint256 buyAmount, uint256 feeAmount, bytes orderUid)
TRADE_TOPIC = "0xa07a543ab8a018198e99ca0184c93fe9050a79400a0a723441f84de1d972cc17"
# Settlement(address indexed solver) — emitted once per settle() by the
# canonical contract, even when the settlement carries no user Trade events
# (a CoW-AMM/JIT rebalance or buffer/interactions-only op). Presence of this
# event from the canonical contract is proof the tx is a real CoW settlement.
# keccak256("Settlement(address)").
SETTLEMENT_EVENT_TOPIC = "0x40338ce1a7c49204f0099533b1e9a7ee0a3d261f84974ab7af36105b8c4e9db4"
# Autopilot reads the auction id from exactly the last 8 bytes of the calldata.
AUCTION_ID_SUFFIX_LEN = 8

# settle(address[] tokens, uint256[] clearingPrices, Trade[] trades,
#        Interaction[][3] interactions)
_TRADE = "(uint256,uint256,address,uint256,uint256,uint32,bytes32,uint256,uint256,uint256,bytes)"
_INTER = "(address,uint256,bytes)"
_SETTLE_TYPES = ["address[]", "uint256[]", f"{_TRADE}[]", f"{_INTER}[][3]"]

# GPv2 order flags: bit 0 = kind (0 sell / 1 buy), bit 1 = partiallyFillable.
FLAG_KIND_BUY = 1
FLAG_PARTIALLY_FILLABLE = 2


def _ceildiv(a: int, b: int) -> int:
    return (a + b - 1) // b


def decode_settlement(calldata_hex: str):
    """Decode a settle() calldata string into tokens[], clearing_prices[],
    trades[] (raw tuples), interactions, and the appended auction_id (last 8
    bytes of the suffix, per autopilot's rule, or None if the prefix does not
    re-encode canonically). Raises ValueError if not a settle() call."""
    if not calldata_hex.startswith("0x"):
        calldata_hex = "0x" + calldata_hex
    if calldata_hex[:10].lower() != SETTLE_SELECTOR:
        raise ValueError(f"not a GPv2 settle() call (selector {calldata_hex[:10]})")
    raw = bytes.fromhex(calldata_hex[10:])
    tokens, prices, trades, interactions = decode(_SETTLE_TYPES, raw)
    canon = encode(_SETTLE_TYPES, (tokens, prices, trades, interactions))
    suffix = raw[len(canon):]
    auction_id = None
    if raw[:len(canon)] == canon and len(suffix) >= AUCTION_ID_SUFFIX_LEN:
        auction_id = int.from_bytes(suffix[-AUCTION_ID_SUFFIX_LEN:], "big")
    return {
        "tokens": [t.lower() for t in tokens],
        "clearing_prices": list(prices),
        "trades": list(trades),
        "interactions": interactions,
        "auction_id": auction_id,
        "suffix_len": len(suffix),
    }


def surplus_from_execution(kind, exec_sell_net, exec_buy, limit_sell, limit_buy):
    """Before-fee surplus of one trade, in SURPLUS-TOKEN atoms, computed from
    the ACTUALLY EXECUTED amounts (the on-chain Trade event) versus the signed
    limit — no clearing prices involved.

    kind='sell': surplus token = BUY token:
        received - ceil(net_sold * limit_buy / limit_sell)
    kind='buy':  surplus token = SELL token:
        floor(received * limit_sell / limit_buy) - net_sold

    This is the un-forgeable form. The earlier price-based reconstruction read a
    clearing-price SLOT the settlement author chooses freely (a settle() may
    carry duplicate token entries; the EVM reads the slot each trade's
    sellTokenIndex/buyTokenIndex points at, which nothing validates for content),
    so a solver could inflate the reconstructed surplus by editing an
    otherwise-unread slot. The executed amounts in the Trade event are what the
    EVM actually moved and cannot be edited after the fact; the signed limits are
    the user's. `exec_sell_net` is the executed sell amount NET of fee (the
    Trade event's sellAmount is fee-inclusive).

    Returns 0 on degenerate inputs; a negative result (limit violation) is
    returned as-is so callers can distinguish infeasible from zero-surplus."""
    if not (limit_sell and limit_buy and exec_sell_net and exec_buy):
        return 0
    if kind == "sell":
        lim_bought = _ceildiv(exec_sell_net * limit_buy, limit_sell)
        return exec_buy - lim_bought
    else:
        lim_sold = exec_buy * limit_sell // limit_buy
        return lim_sold - exec_sell_net


def surplus_token(kind, sell_token, buy_token):
    """Official surplus token: buy token for sell orders, sell token for buy."""
    return buy_token if kind == "sell" else sell_token


def to_native(atoms, ref_price):
    """Convert surplus-token atoms to native wei at the auction referencePrice."""
    if ref_price is None:
        return None
    return atoms * ref_price // 10 ** 18


def trade_events(logs):
    """Extract executed amounts + orderUid from GPv2 Trade events — ground truth.

    Filters by BOTH topic0 and the emitting address (the canonical settlement
    contract), so a foreign contract emitting a same-signature event can't shift
    the positional events[i] <-> trades[i] alignment."""
    out = []
    for lg in logs:
        topics = lg.get("topics") or []
        if not topics or topics[0].lower() != TRADE_TOPIC:
            continue
        if lg.get("address", "").lower() != SETTLEMENT:
            continue
        d = (lg.get("data") or "")[2:]
        if len(d) < 560:  # 5 words + offset + length + 56-byte uid
            continue
        out.append({
            "owner": ("0x" + lg["topics"][1][-40:]).lower(),
            "sell_token": ("0x" + d[24:64]).lower(),
            "buy_token": ("0x" + d[64 + 24:128]).lower(),
            "sell_amount": int(d[128:192], 16),
            "buy_amount": int(d[192:256], 16),
            "uid": ("0x" + d[448:560]).lower(),
        })
    return out


def selftest():
    """Deterministic surplus-math vectors — both order kinds and boundaries."""
    # sell: sold 100 net, received 200, limit 100->150 => lim_bought=150 => +50
    assert surplus_from_execution("sell", 100, 200, 100, 150) == 50
    # sell exactly at limit => 0
    assert surplus_from_execution("sell", 100, 150, 100, 150) == 0
    # sell below limit (infeasible) => negative
    assert surplus_from_execution("sell", 100, 100, 100, 150) < 0
    # buy: received 100, paid 150 net, limit 200->100 => lim_sold=200 => +50
    assert surplus_from_execution("buy", 150, 100, 200, 100) == 50
    # buy exactly at limit => 0
    assert surplus_from_execution("buy", 200, 100, 200, 100) == 0
    assert surplus_from_execution("sell", 0, 200, 100, 150) == 0
    assert surplus_token("sell", "a", "b") == "b"
    assert surplus_token("buy", "a", "b") == "a"
    return True
